pkg/util/timeout.go

```diff
@@ -14,29 +14,38 @@ type TimeoutError struct {
 // Error implements the Go error interface.
 func (t *TimeoutError) Error() string {
 	if len(t.message) > 0 {
-		return fmt.Sprintf(t.message, t.after)
+		return fmt.Sprintf("%s timed out after %v", t.message, t.after)
 	}
-	return fmt.Sprintf("calling the function timeout after %v", t.after)
+	return fmt.Sprintf("function timed out after %v", t.after)
 }
 
-// TimeoutAfter executes the provide function and return the TimeoutError in
-// case when the execution time of the provided function is bigger than provided
-// time duration.
-func TimeoutAfter(t time.Duration, errorMsg string, fn func() error) error {
+// TimeoutAfter executes the provided function and returns TimeoutError in the
+// case that the execution time of the function exceeded the provided duration.
+// The provided function is passed the timer in case it wishes to reset it.  If
+// so, the following pattern must be used:
+// if !timer.Stop() {
+//   return &TimeoutError{}
+// }
+// timer.Reset(timeout)
+func TimeoutAfter(t time.Duration, errorMsg string, f func(*time.Timer) error) error {
 	c := make(chan error, 1)
+	timer := time.NewTimer(t)
 	go func() {
-		defer close(c)
-		c <- fn()
+		err := f(timer)
+		if !IsTimeoutError(err) {
+			c <- err
+		}
```

Comment:
why wouldn't you send the timeout err into the channel too, does it really matter which way the error bubbles up?